implementations/rust/ockam/ockam_core/src/flow_control/flow_controls/tests.rs

```diff
@@ -0,0 +1,209 @@
+                    );
+                    vec![address]
+                }
+            }
+        }
+        _ => panic!(),
+    }
+}
+
+fn add_random_spawner(flow_controls: &FlowControls, mut rng: &mut ThreadRng) -> Vec<Address> {
+    let address = Address::random_local();
+
+    let choice: f32 = rng.gen();
+
+    if choice < 0.8 {
+        // Add a Spawner to a new FlowControlId
+        let flow_control_id = FlowControls::generate_id();
+
+        flow_controls.add_spawner(address.clone(), &flow_control_id);
+
+        vec![address]
+    } else {
+        // Add a Spawner to an existing FlowControlId
+        let spawner_flow_control_id = flow_controls
+            .spawners
+            .read()
+            .unwrap()
+            .values()
+            .choose(&mut rng)
+            .cloned();
+
+        match spawner_flow_control_id {
+            Some(spawner_flow_control_id) => {
+                flow_controls.add_spawner(address.clone(), &spawner_flow_control_id);
+                vec![address]
+            }
+            _ => vec![],
+        }
+    }
+}
+
+fn add_random_producer(flow_controls: &FlowControls, mut rng: &mut ThreadRng) -> Vec<Address> {
+    let address = Address::random_local();
+
+    let choice: f32 = rng.gen();
+
+    if choice < 0.5 {
+        // Add a Producer without a Spawner
+        let flow_control_id = FlowControls::generate_id();
+
+        flow_controls.add_producer(address.clone(), &flow_control_id, None, vec![]);
+
+        vec![address]
+    } else {
+        // Add a Producer with a Spawner
+        let flow_control_id = FlowControls::generate_id();
+
+        let spawner_flow_control_id = flow_controls
+            .spawners
+            .read()
+            .unwrap()
+            .values()
+            .choose(&mut rng)
+            .cloned();
+
+        match spawner_flow_control_id {
+            Some(spawner_flow_control_id) => {
+                let choice2: f32 = rng.gen();
+                if choice2 < 0.5 {
+                    // No additional address
+                    flow_controls.add_producer(
+                        address.clone(),
+                        &flow_control_id,
+                        Some(&spawner_flow_control_id),
+                        vec![],
+                    );
+                    vec![address]
+                } else {
+                    // Random additional address
+                    let additional_address = Address::random_local();
+                    flow_controls.add_producer(
+                        address.clone(),
+                        &flow_control_id,
+                        Some(&spawner_flow_control_id),
+                        vec![additional_address.clone()],
+                    );
+                    vec![address, additional_address]
+                }
+            }
+            None => vec![],
+        }
+    }
+}
+
+#[test]
+fn test_cleanup() {
+    let mut rng = rand::thread_rng();
```

Comment:
When testing with a random factor, it's simpler to debug failure when either we have a fixed seed or the seed is printed out

Reply:
I think there is a benefit of having a random seed each time because there is a possible situation where the test succeeds for the fixed seed and doesn't succeed for some other seed. As for seeding - AFAIK, we don't want tests to print anything.

Reply:
In my previous company we had the following strategy:

 - always run the regular CI with a fixed seed because we value non-flakiness more than finding an exceptional use case. Especially because in many cases a failing random tests indicated an incorrect assumption of the test assertion but the production code was fine
 
 - run property tests in a weekly CI with a random seed and more tests per property (BTW @SanjoDeundiak why didn't you use `quickcheck` there instead of hand-rolling arbitrary data generation?)

I think that it is a good balance between stability of the dev process and bug finding